internal/pkg/agent/application/coordinator/coordinator.go

```diff
@@ -656,77 +658,146 @@ func (c *Coordinator) watchRuntimeComponents(ctx context.Context) {
 	// and just idle until Coordinator shuts down.
 	if c.runtimeMgr != nil {
 		subChan = c.runtimeMgr.SubscribeAll(ctx).Ch()
 	}
+	if c.otelMgr != nil {
+		otelChan = c.otelMgr.Watch()
+	}
 	for {
 		select {
 		case <-ctx.Done():
 			return
-		case s := <-subChan:
-			oldState, ok := state[s.Component.ID]
-			if !ok {
-				componentLog := coordinatorComponentLog{
-					ID:    s.Component.ID,
-					State: s.State.State.String(),
-				}
-				logBasedOnState(c.logger, s.State.State, fmt.Sprintf("Spawned new component %s: %s", s.Component.ID, s.State.Message), "component", componentLog)
-				for ui, us := range s.State.Units {
-					unitLog := coordinatorUnitLog{
-						ID:    ui.UnitID,
-						Type:  ui.UnitType.String(),
-						State: us.State.String(),
-					}
-					logBasedOnState(c.logger, us.State, fmt.Sprintf("Spawned new unit %s: %s", ui.UnitID, us.Message), "component", componentLog, "unit", unitLog)
-				}
-			} else {
-				componentLog := coordinatorComponentLog{
-					ID:    s.Component.ID,
-					State: s.State.State.String(),
-				}
-				if oldState.State != s.State.State {
-					cl := coordinatorComponentLog{
-						ID:       s.Component.ID,
-						State:    s.State.State.String(),
-						OldState: oldState.State.String(),
-					}
-					logBasedOnState(c.logger, s.State.State, fmt.Sprintf("Component state changed %s (%s->%s): %s", s.Component.ID, oldState.State.String(), s.State.State.String(), s.State.Message), "component", cl)
-				}
-				for ui, us := range s.State.Units {
-					oldUS, ok := oldState.Units[ui]
-					if !ok {
-						unitLog := coordinatorUnitLog{
-							ID:    ui.UnitID,
-							Type:  ui.UnitType.String(),
-							State: us.State.String(),
-						}
-						logBasedOnState(c.logger, us.State, fmt.Sprintf("Spawned new unit %s: %s", ui.UnitID, us.Message), "component", componentLog, "unit", unitLog)
-					} else if oldUS.State != us.State {
-						unitLog := coordinatorUnitLog{
-							ID:       ui.UnitID,
-							Type:     ui.UnitType.String(),
-							State:    us.State.String(),
-							OldState: oldUS.State.String(),
-						}
-						logBasedOnState(c.logger, us.State, fmt.Sprintf("Unit state changed %s (%s->%s): %s", ui.UnitID, oldUS.State.String(), us.State.String(), us.Message), "component", componentLog, "unit", unitLog)
-					}
-				}
-			}
-			state[s.Component.ID] = s.State
-			if s.State.State == client.UnitStateStopped {
-				delete(state, s.Component.ID)
-			}
+		case componentState := <-subChan:
+			logComponentStateChange(c.logger, state, &componentState)
 			// Forward the final changes back to Coordinator, unless our context
 			// has ended.
 			select {
-			case c.managerChans.runtimeManagerUpdate <- s:
+			case c.managerChans.runtimeManagerUpdate <- componentState:
 			case <-ctx.Done():
 				return
 			}
+		case otelStatus := <-otelChan:
+			componentStates, err := translate.GetAllComponentStates(otelStatus, c.componentModel)
+			if err != nil {
+				c.setOTelError(err)
```

Comment:
This feels like a bug reading the code. I read the code to understand below that `componentStates` can be nil and it will not cause an issue. I think it would be best to add a comment to explain this. That makes it clearer when reading the code that continuing here is okay.